marker_api/src/context/map.rs

```diff
@@ -0,0 +1,90 @@
+use crate::{
+    ast::{
+        expr::ExprKind,
+        item::{Body, EnumVariant, Field, ItemKind},
+        stmt::StmtKind,
+        BodyId, ExprId, FieldId, ItemId, StmtId, VariantId,
+    },
+    ffi,
+    lint::{Level, Lint},
+    prelude::{HasNodeId, NodeId},
+};
+
+/// A map, which allows the request of AST nodes by their ids.
+///
+/// An instance of this map can be accessed from [`MarkerContext::ast`](super::MarkerContext::ast).
+#[repr(C)]
+#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
+pub struct AstMap<'ast> {
+    callbacks: AstMapCallbacks<'ast>,
+}
+
+impl<'ast> AstMap<'ast> {
+    pub fn lint_level_at(&self, lint: &'static Lint, node: impl HasNodeId) -> Level {
+        (self.callbacks.lint_level_at)(self.callbacks.data, lint, node.node_id())
+    }
+
+    /// Returns the [`ItemKind`] belonging to the given [`ItemId`].
+    pub fn item(&self, id: ItemId) -> Option<ItemKind<'ast>> {
+        (self.callbacks.item)(self.callbacks.data, id).copy()
+    }
+
+    /// Returns the [`EnumVariant`] belonging to the given [`VariantId`].
+    ///
+    /// This might return [`None`] if the requested [`EnumVariant`] is no longer
+    /// available. This can happen, if the [`EnumVariant`] comes from a different
+    /// crate which is currently not loaded or if the [`VariantId`] has become
+    /// invalid.
+    pub fn variant(&self, id: VariantId) -> Option<&EnumVariant<'ast>> {
```

Comment:
All these methods return an `Option`. I suppose the option will be `None` in case of some really hard-to-reproduce bug. I suppose if the `VariantId` was obtained from the current context and wasn't read from some persistent storage or wasn't constructed manually using `unsafe` then this will never return `None`. Maybe it makes sense then to just panic in case `None` is returned complaining that `VariantId` is now dangling and your code contains a bug.

Otherwise what would the use of `marker_api` do when it sees an option here? Is there any possible use case where the user of these method won't unwrap the option? I doubt someone ever needs to check that the option is `None`, and do something else meaningful.

Reply:
I had three reasons:
1. If a user stores the *ID* somewhere and then reloads it or constructs an invalid one
    * I didn't consider that a user would actually not be able to create an ID, without unsafe code :sweat_smile:
2. Rustc compiles each crate individually. During linting it's possible to get the `ItemId` of an item from a dependency (For example by resolving the target of a function call). However, there is no AST representation of the item available, since it's not in the currently compiled crate.
3. If we ever support other drivers, like IDE's where nodes might be deleted again.
    * This one is the least important, as compilers are currently the main target of Marker. And the IDs already have a note in their documentation, that they're not allowed to be stored across `check_*` function boundaries. I considered giving them an `'ast` lifetime, but that sounds like a lot of hassle.

I think 2. is the main reason to keep the result wrapped in an `Option`. We could remove it from the `expr` method, since it's not possible to get an expression id from external expressions.

I'll also add more documentation to `AstMap` to describe when these methods can return `None`.

Rustc uses a general `find() -> Option<Node>` function and then specific `expect_expr()` functions. That could be a model we can easily adapt :thinking: 

Does this reasoning make sense?

Reply:
So, for example, if we lint crate `foo`, then how can we get the `VariantId` from say `std` crate without unsafe read from external storage?

---

Yeah, I agree that for the general use case when users don't try to access the info from the external crate we should make that more convenient to obtain that info (panic by default). I'd make the functions like these targeting for the optimistic case (when users don't access AST from external crates, if that's even possible):
```rust
// Panics (general use case)
fn variant(&self, id: VariantId) -> &EnumVariant<'ast>;

// Returns `None`, but I wonder if this will ever be useful to anyone
fn try_variant(&self, id: VariantId) -> Option<&EnumVariant<'ast>>;
```

Reply:
Let's say we have the following input:

```
fn main() {
	let x = Some(12)
}
``` 

Then the lint could do this:

```
fn check_stmt(...) {
    if let StmtKind::Let(local) = stmt {
        if let Some(ExprKind::Ctor(ctor)) = local.init() {
            // https://docs.rs/marker_api/latest/marker_api/ast/struct.AstQPath.html
            let path = ctor.path();

            if let AstPathTarget::Variant(var_id) = path.resolve() {
                // Here we have a the valid `VariantId` ID for `Some`.
                //
                // Right now, it would be useless, but at one point, Marker should enable
                // users to request IDs from dependencies and get the semantic variant
                // information, which is actually available
            }
        }
    }
}
```

Having the `try_` prefix might work well :+1:

Reply:
I've decided to use the `expect_` prefix instead. IMO it's too easy to get a valid ID for an unavailable AST node, as to make the panicking variant the default.

But there are some nodes, where it's certain, that a given ID also has an AST node. This being bodies, statements, and expressions. For those, I've changed the return type. Beforehand, I was a bit uncertain about this. Statements, for example, can also wrap items, but my current understanding is that this should still be safe.

I'll make the adjustments and then run it on a few test crates, like tokio and bevy.

Reply:
I'm not sure why `rustc` uses `expect_` prefix, because if you see `expect` you'd likely think there will be an argument for the expect message, but there isn't any.

For example, `Result` uses `unwrap_` prefix for [`unwrap_err`](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.unwrap_err), and it seems more logical to me